contents/java/Spring.md

```diff
@@ -0,0 +1,291 @@
+pluggable configurations. 
+
+For example, say we wish to log something into the console everytime before we call the method `getEmployeeById`.
+We can then write aspect class annotated with `@Aspect` annotation and write point-cut expressions to log program information from
+<tooltip content="Join point is a point of execution of the program, such as the execution of a method or the handling of an exception.">
+<i>joint-point</i>
+</tooltip> methods.
+```java
+@Aspect
+public class EmployeeCRUDAspect {
+      
+    @Before("execution(* EmployeeManager.getEmployeeById(..))") // point-cut expressions
+    public void logBefore(JoinPoint joinPoint) {
+        System.out.println("EmployeeCRUDAspect.logBefore() : "
+                + joinPoint.getSignature().getName());
+    }
+    
+    @After("execution(* EmployeeManager.getEmployeeById(..))") // point-cut expressions
+    public void logAfter(JoinPoint joinPoint) {
+        System.out.println("EmployeeCRUDAspect.logAfter() : "
+                + joinPoint.getSignature().getName());
+    }
+}
+```
+    
+And the corresponding `EmployeeManager.java` file could be:
+```java
+@Component
+public class EmployeeManager
+{
+    public EmployeeDTO getEmployeeById(Integer employeeId) {
+        System.out.println("Method getEmployeeById() called");
+        return new EmployeeDTO();
+    }
+}
+```
+In the above example, `logBefore()` will be executed before the `getEmployeeById()` method because it matches the join point specified earlier.
+Similarly, `logAfter()` will be executed after `getEmployeeById()`.
+The output of the above code will be:
+```
+EmployeeCRUDAspect.logBefore() : getEmployeeById
+Method getEmployeeById() called
+EmployeeCRUDAspect.logAfter() : getEmployeeById
+```
+
+The declaration of aspects and
+<tooltip content="Advice is an action taken by an aspect at a particular join point.">
+<i>advices </i>
+</tooltip>could also be written directly in the XML templates. This makes the code more readable and
+also allows for easier maintainance as we can add/remove concerns 
+by changing XML configuration file without recompiling source code.
+
+For example, the above configuration could be written in a XML file as:
+```xml
+ <aop:config>
+    <!-- Spring AOP Pointcut definitions -->
+    <aop:pointcut id="CRUDOperation" expression="execution(* EmployeeManager.*(..))" />
+ 
+    <!-- Spring AOP aspect 1 -->
+    <aop:aspect id="EmployeeCRUD" ref="crudAspectBean">
+             
+        <!-- Spring AOP advises -->
+        <aop:before pointcut-ref="CRUDOperation" method="logBefore" />
+        <aop:after pointcut-ref="CRUDOperation" method="logAfter" />
+             
+    </aop:aspect>
+
+    <!-- Spring AOP aspect instances -->
+    <bean id="crudAspectBean" class="com.aop.EmployeeCRUDAspect" />
+     
+    <!-- Target Object -->
+    <bean id="employeeManager" class="com.aop.EmployeeManagerImpl" />
+</aop:config>
+```
```

Comment:
We should not write as if AOP is a Spring thing. Rather, Spring is _using_ AOP.
Given that, try not to go too much into AOP. If anything, it should be separate chapter by itself.